### PR TITLE
[vcpkg] Add -isysroot args to CPPFLAGS

### PR DESCRIPTION
**Pull request by m-kuhn**
So preprocessor only configure tests get the right includes

**Describe the pull request**

- #### What does your PR fix?
  `CPPFLAGS` are currently lacking `-isysroot` when invoking autoconf tools. This leads to preprocessor only tests missing headers and producing bogus configuration results. This is an issue on macos where system headers do not live in the system root but rather a dedicated sdk folder.

For more information see https://github.com/microsoft/vcpkg/pull/21847/files#r776056939

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  all

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  No port touched

### CODE REVIEW

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 1]
```diff
@@ -103,6 +103,17 @@ macro(z_vcpkg_extract_cpp_flags_and_set_cflags_and_cxxflags flag_suffix)
     string(STRIP "${CPPFLAGS_${flag_suffix}}" CPPFLAGS_${flag_suffix})
     string(STRIP "${CFLAGS_${flag_suffix}}" CFLAGS_${flag_suffix})
     string(STRIP "${CXXFLAGS_${flag_suffix}}" CXXFLAGS_${flag_suffix})
+
+    # Extract -isysroot from CFLAGS and add it to CPPFLAGS
+    # so that preprocessor-only configure tests get the right includes
+    separate_arguments(C_FLAGS NATIVE_COMMAND PROGRAM SEPARATE_ARGS "${VCPKG_DETECTED_CMAKE_C_COMPILER} ${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_make.cmake:109]:
```suggestion
    separate_arguments(c_flags_sep NATIVE_COMMAND "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
```

Why do you need the special PROGRAM option with SEPARATE_ARGS here?

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 2]
```diff
@@ -103,6 +103,17 @@ macro(z_vcpkg_extract_cpp_flags_and_set_cflags_and_cxxflags flag_suffix)
     string(STRIP "${CPPFLAGS_${flag_suffix}}" CPPFLAGS_${flag_suffix})
     string(STRIP "${CFLAGS_${flag_suffix}}" CFLAGS_${flag_suffix})
     string(STRIP "${CXXFLAGS_${flag_suffix}}" CXXFLAGS_${flag_suffix})
+
+    # Extract -isysroot from CFLAGS and add it to CPPFLAGS
+    # so that preprocessor-only configure tests get the right includes
+    separate_arguments(C_FLAGS NATIVE_COMMAND PROGRAM SEPARATE_ARGS "${VCPKG_DETECTED_CMAKE_C_COMPILER} ${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
+    foreach(C_FLAG ${C_FLAGS})
+        if(${LAST_C_FLAG} MATCHES "-isysroot")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_make.cmake:111]:
```suggestion
        if("${LAST_C_FLAG}" STREQUAL "-isysroot")
```

File: scripts/cmake/vcpkg_configure_make.cmake (hunk 1 above)
Comment (m-kuhn) [reply scripts/cmake/vcpkg_configure_make.cmake:109]: I may have been mislead by the examples on the docs page... Will try a simpler construct.

File: scripts/cmake/vcpkg_configure_make.cmake (hunk 2 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_make.cmake:111]: Ich feel like this should use `list(FIND <list> <value> <output variable>)` together with `list(SUBLIST <list> <begin> <length> <output variable>)`

and also remove `isysroot` from the normal `C(XX)_FLAGS`

Comment (m-kuhn) [reply scripts/cmake/vcpkg_configure_make.cmake:111]: 1. sounds good
2. was implemented analogue to https://mail.gnome.org/archives/commits-list/2014-February/msg09078.html but I share your feeling

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 3]
```diff
@@ -92,6 +92,19 @@ macro(z_vcpkg_extract_cpp_flags_and_set_cflags_and_cxxflags flag_suffix)
         set(CFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
         set(CXXFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${flag_suffix}}")
     endif()
+
+    # Extract -isysroot from CFLAGS so that preprocessor-only configure tests get the right includes.
+    set(C_FLAGS "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
+    separate_arguments(C_FLAGS)
+    list(FIND C_FLAGS "-isysroot" ISYSROOT_FLAG_POS)
```
Comment (JackBoosY) [inline scripts/cmake/vcpkg_configure_make.cmake:99]: Why not use `if ("${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}" MATCHES "-isysroot")`?
Are there other similar options?

Comment (m-kuhn) [reply scripts/cmake/vcpkg_configure_make.cmake:99]: `${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}` is a string with all flags, separate_arguments splits it into a list and this line finds the index in this list to extract just this argument, so we can use it. I don't see how `MATCHES` could replace this logic.
Am I missing the point?

Comment (dg0yt) [reply scripts/cmake/vcpkg_configure_make.cmake:99]: I think the irritating point is that one part of flag processing is done in with regex on the string, and one part of flag processing is (now) done on a list.

Comment (m-kuhn) [reply scripts/cmake/vcpkg_configure_make.cmake:99]: The regex approach will likely run into limitations if the path contains spaces

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 4]
```diff
@@ -92,6 +92,19 @@ macro(z_vcpkg_extract_cpp_flags_and_set_cflags_and_cxxflags flag_suffix)
...
+
+    # Extract -isysroot from CFLAGS so that preprocessor-only configure tests get the right includes.
+    set(C_FLAGS "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
+    separate_arguments(C_FLAGS)
+    list(FIND C_FLAGS "-isysroot" ISYSROOT_FLAG_POS)
+    if(NOT ISYSROOT_FLAG_POS EQUAL -1)
+        MATH(EXPR ISYSROOT_FLAG_POS "${ISYSROOT_FLAG_POS}+1")
+        list(SUBLIST C_FLAGS ${ISYSROOT_FLAG_POS} 1 ISYSROOT)
+        set(CPPFLAGS_${flag_suffix} "${CPPFLAGS_${flag_suffix}} -isysroot ${ISYSROOT}")
+        string(REGEX REPLACE "-isysroot ${ISYSROOT}" "" CFLAGS_${flag_suffix} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${flag_suffix}}")
```
[the remark this answers is not in the document]
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_make.cmake]: ? 

Why and where? 

It should always be `$(CPPFLAGS) $(CFLAGS)` in the makefiles.

[the remark this answers is not in the document]
Comment (m-kuhn) [reply scripts/cmake/vcpkg_configure_make.cmake]: From `gettext-runtime/configure` in `checking whether <wchar.h> uses 'inline' correctly`:

```
if $CC -o conftest$ac_exeext $CFLAGS $LDFLAGS conftest1.$ac_objext conftest2.$ac_objext $LIBS >&5 2>&1; then
```

**Comment by BillyONeal** [comment]
> This is an issue on macos where system headers do not live in the system root but rather a dedicated sdk folder.

Shouldn't the toolchain be making sure any such SDK folder is available rather than having each build system under the sun try to fix it up like this?

**Comment by dg0yt** [comment]
> > This is an issue on macos where system headers do not live in the system root but rather a dedicated sdk folder.
> 
> Shouldn't the toolchain be making sure any such SDK folder is available rather than having each build system under the sun try to fix it up like this?

In vcpkg, "each build system under the sun" should take the details from CMake. Taking from CMake is done here:
https://github.com/microsoft/vcpkg/blob/8e1f46de11423e5f7d2395458f5252738ffffbbd/scripts/get_cmake_vars/CMakeLists.txt#L101-L114

Should it be taken differently, before transforming it for a particular build system such as autotools?

**Comment by BillyONeal** [comment]
> In vcpkg, "each build system under the sun" should take the details from CMake. Taking from CMake is done here:
> 
> https://github.com/microsoft/vcpkg/blob/8e1f46de11423e5f7d2395458f5252738ffffbbd/scripts/get_cmake_vars/CMakeLists.txt#L101-L114
> 
> Should it be taken differently, before transforming it for a particular build system such as autotools?

It seems like this should mean `-isysroot` should already be present in `VCPKG_DETECTED_Xxx`?

**Comment by BillyONeal** [comment]
> it is the problem is that a pure preprocessor run does not have the flag.

Shouldn't a pure preprocessor run have all flags though? This is far from the only compiler switch that may affect preprocessor results.

**Comment by dg0yt** [comment]
> > Should it be taken differently, before transforming it for a particular build system such as autotools?
> 
> It seems like this should mean `-isysroot` should already be present in `VCPKG_DETECTED_Xxx`?

Well, there is no `VCPKG_DETECTED_Xxx` for the preprocessor. There is C and CXX. It somewhat questionable that we first push the sysroot flag into the C/CXX flags, and later pull it out again, instead of forwarding it separately.

**Comment by m-kuhn** [comment]
> The question here is rather why does it only do you preprocessor check and no compile check. Maybe availability of includes is only checked with the preprocessor?

Availability of certain contents of includes in this case, FWIW.

```sh
# checking for uid_t in sys/types.h
  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
/* end confdefs.h.  */
#include <sys/types.h>

_ACEOF
if (eval "$ac_cpp conftest.$ac_ext") 2>&5 |
  $EGREP "uid_t" >/dev/null 2>&1; then :
  ac_cv_type_uid_t=yes
else
  ac_cv_type_uid_t=no
fi
```

But I'm not sure what this knowledge changes, properly populating `CPPFLAGS` is probably the more portable way forward than changing such constructs in various autoconf ports?

**Comment by m-kuhn** [comment]
> > > Should it be taken differently, before transforming it for a particular build system such as autotools?
> > 
> > 
> > It seems like this should mean `-isysroot` should already be present in `VCPKG_DETECTED_Xxx`?
> 
> Well, there is no `VCPKG_DETECTED_Xxx` for the preprocessor. There is C and CXX. It somewhat questionable that we first push the sysroot flag into the C/CXX flags, and later pull it out again, instead of forwarding it separately.

Adding something like `VCPKG_DETECTED_CPP_FLAGS` would be somewhat alien because there is no cmake equivalent. But overall it might still be cleaner.

**Comment by m-kuhn** [comment]
@JackBoosY is the answer good enough to remove author-response?

**Comment by JackBoosY** [comment]
Ping @BillyONeal @dg0yt @Neumann-A for review agian.

**Comment by BillyONeal** [comment]
> > It seems like this should mean -isysroot should already be present in VCPKG_DETECTED_Xxx?
> 
> it is the problem is that a pure preprocessor run does not have the flag.

It doesn't look like this is setting a pure preprocessor variable though, it's setting CPPFLAGS.

> Adding something like `VCPKG_DETECTED_CPP_FLAGS` would be somewhat alien because there is no cmake equivalent. But overall it might still be cleaner.

I'm confused, why is that not VCPKG_DETECTED_CXX_FLAGS as already set here?

https://github.com/microsoft/vcpkg/pull/25201/commits/4b2916b698e2c61c986a051da47956d3fee6f83d#diff-b92b07d5c0c681687d71463c6a368fae98e9ffedb241e2aadbde3aac688899a9R93

**Comment by m-kuhn** [comment]
Sorry if I'm stating something obvious, just to avoid a potential source of confusion:

CFLAGS: C Flags
CXXFLAGS: C++ Flags
CPPFLAGS: Preprocessor Flags (a concept not used natively in cmake, where they are added to CFLAGS/CXXFLAGS)

My understanding:
For running `configure` we want `CPPFLAGS` to be set because some checks (apparently) rely on preprocessor runs.
Since cmake internally doesn't separate between pre-processor and processor flags, there's no simple cmake variable we could extract for this purpose and we need to either rely on heuristics (extracting CPPFLAGS via regex magic for `defines`, `-isysroot`, other keywords) or populate this in the toolchain file (although I'm not sure this is the only point of entry for these).

Currently this is being completely ignored and things work because on osx `-isysroot` is implicitly added by a wrapper executable called `gcc` that internally calls `clang` with `-isysroot` already populated and the sysroot on linux builds is often left to default. This is a potential source of problems, not least for cross compiling (android, ios) also.

**Comment by BillyONeal** [comment]
Ah, CPP as in C PreProcessor, not C PlusPlus. No, that was not obvious to me. :) My "coming from a platform where the preprocessor is not a separate program" is showing.

**Comment by m-kuhn** [comment]
https://github.com/microsoft/vcpkg/pull/25670 implements the same approach as for meson, discarding this approach until the other (osx specific approach) runs into limitations

**Pull request closed without merging**